### PR TITLE
[Specification] CDS-WG1-01 - Server Metadata

### PR DESCRIPTION
**Pull request by daniel-utilityapi**
This is the initial draft of a proposed specification for utility servers to publish metadata about their details and capabilities.

Relevant Issue: https://github.com/lfe-cds/CDS-Registration/issues/2

**NOTE: This is a draft in progress (see "TODO" throughout the draft) and will continue to be developed and modified based on feedback and comments in this pull request and during meetings**

See a rendered preview of this specification on the maintainer's branch:
https://daniel-utilityapi.github.io/CDS-Registration/specs/cds-wg1-01

## Timeline

* [x] This pull request has undergone significant review by working group participants (2024-2025)
* [x] It is ready for a call to merge as a DRAFT version on the main CDS Registration working group website (June 2025)
  * Deadline for feedback before merging: **August 7, 2025**

Further edits can be managed via individual pull request (Q3 2025). When the draft on the main CDS website is considered consensus, it can be sent to the LFESS steering committee as final (target Q4 2025)

### CODE REVIEW

**Comment by daniel-utilityapi** [comment]
## NOTE: This pull request was migrated from https://github.com/carbon-data-specification/Customer-Data/pull/23

Previous comments and feedback have been manually copied over by the maintainer from the previous pull request.

**Comment by daniel-utilityapi** [comment]
From @eloifabrega ([on April 13, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#issuecomment-1506672798)):

---
Hello!

I was reviewing this and overall looks good to me. Nevertheless, I have a couple of comments:

- Would it be useful to specify here more characteristics of ¨commodity_type¨ data? Specifically for electricity, which data does the server offer or can be accessed? Billing, interval? Production, consumption? Maybe then they will not have to check the documentation linked for a specific utility, and this could be crucial for some Energy Service Providers that need to access more granular data for example. 

- Also adding in server capabilities, might be useful (for an extended specifications phase) to add if the server allows access to third parties (and potential costs for that). Unless I miss it, this could only be checked in the ¨documentation¨ and from my experience, is not always easy to find from the potential servers (utilities). 

Hope this can help and I am happy to discuss this all together!
Many thanks :)
---

**Comment by daniel-utilityapi** [comment]
From @greg-flexidao ([on April 13, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#issuecomment-1507149600)):

---
Hi All,
Sorry I didn't yet check in depth the rest of the spec, my initial suggestions would be:
* Consider date-time format according to RFC-3339 in order to explicitly define the time of the event/period. The example in the specification provides explicit time (ISO8601), note that it is also RFC-3339 compliant. ISO unfortunately also allows for ambiguous datetimes that are one of the main plagues and source of issues in the industry. See https://ijmacd.github.io/rfc3339-iso8601/ for comparison.
* I would say that the endpoint/api should in general follow REST API principles. Unfortunately there is no ultimate source of truth for standard REST APIs (that I know of), here is however a comprehensive guide that could be followed and it can be referenced in the standard, that also includes the cases that have been mentioned. https://restfulapi.net/

Happy to participate in further spec development! 
---

**Comment by daniel-utilityapi** [comment]
From @halliecramer1 ([on April 14, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#pullrequestreview-1386252384)):

---

Added some thoughts to address questions raised in Thursday meeting

---

File: specifications/cdsc-wg1-01.md [hunk 1]
```diff
@@ -1 +1,292 @@
...
+* `documentation` - _URL_ - (REQUIRED) Where Clients can find the utility or entity's technical reference documentation
+* `support` - _URL_ - (REQUIRED) Where Clients can find contact information for technical support on the Server's capabilities
+* `infrastructure_types` - _Array[[InfrastructureType](#infrastructure-types)]_ - (REQUIRED) A list of infrastructure categories that are applicable to this utility or entity, which may be an empty array if the Server is not any of the possible infrastructure types
+* `commodity_types` - _Array[[CommodityType](#commodity-types)]_ - (REQUIRED) A list of services or commodities that are applicable to this utility or entity, which may be an empty array if the Server does not have an applicable commodity
+* `capabilities` - _Array[[ServerCapability](#server-capabilities)]_ - (REQUIRED) What capabilities and functionality this Server supports, which may be an empty array if the Server is not offering any additional capabilities beyond just this metadata object
+* `coverage` - _URL_ - (OPTIONAL) Where to find the [Coverage Endpoint](#coverage-endpoint)
```
Comment (daniel-utilityapi) [inline specifications/cdsc-wg1-01.md:88]: From @halliecramer1 ([on April 14, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#discussion_r1167326314)):

---

Suggest restructuring coverage - instead of as a capability, create a `coverage_types` array, and for each coverage type in that list, you have a (nested?) entry format. For example, `territory` is one coverage type, and then you nest the Coverage Entry Format - which could have fields specific to that type - but `sub_type` would replace `type` and refer to the service type (e.g. `electric_service`) instead of to territory as a type. Other coverage types could include: `customer_class`, and/or `rate_class` where `sub_type` is e.g. residential / commercial / industrial / client (in case of `energy_solutions_provider`), and/or the specific tariff name. 

Then for each capability, you can include a list of coverage subtypes that it applies to (or list of type, subtype key-value pairs).

---

File: specifications/cdsc-wg1-01.md [hunk 2]
```diff
@@ -1 +1,292 @@
...
+
+### 3.4. Infrastructure Types <a id="infrastructure-types" href="#infrastructure-types" class="permalink">🔗</a>
+
+Infrastructure types are general categories of utility and energy grid entities that are likely to need to provide access to metadata and capabilities under this specification. The following list of strings are an enumerated set of infrastructure types that MAY be included in the `infrastructure_types` array of values in the metadata object.
+
+* `distribution_utility` - A electric, natural gas, and/or water utility that distributes the utility service to the end customer (e.g. "who maintains the wires to your house")
+* `metering_provider` - An entity that maintains the utility metering infrastructure and/or reads end customer meters (e.g. "who reads your meter")
+* `supplier` - An energy supplier for the end customer (e.g. "who you buy power from")
+* `central_repository` - A government mandated entity that collects customer usage information into a centralized repository for re-distribution (e.g. a state-level meter data warehouse for deregulated jurisdictions)
+
```
Comment (daniel-utilityapi) [inline specifications/cdsc-wg1-01.md:109]: From @halliecramer1 ([on April 14, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#discussion_r1167327730)):

---

Add `energy_solutions_provider`, e.g. a Scheduler in the U.S. or a Balancing Responsible Party in Europe, who can provide data on specific transmission level generation assets

---

File: specifications/cdsc-wg1-01.md [hunk 3]
```diff
@@ -1 +1,292 @@
...
+Servers MAY additionally support other public and restricted URLs for special case or private metadata endpoints, as needed.
+
+### 3.2. Metadata Object Format <a id="metadata-object-format" href="#metadata-object-format" class="permalink">🔗</a>
+
+Metadata objects are formatted as JSON objects and contain named values, some of which are required and some of which are optional. The following values are included in the default list available in metadata objects.
+
+* `cds_metadata_version` - _string_ - (REQUIRED) The version of metadata object format this Server follows, which for this version of the specification is `v1`
+* `cds_metadata_url` - _string_ - (REQUIRED) A URL link to this metadata object's canonical network location
+* `created` - _ISO8601 datetime_ - (REQUIRED) When the metadata url was first made available
+* `updated` - _ISO8601 datetime_ - (REQUIRED) When this metadata object was last modified
```
Comment (daniel-utilityapi) [inline specifications/cdsc-wg1-01.md:79]: From @jhughes22 ([on May 25, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#discussion_r1205737110)):

---

Are zoned times allowed or are timestamps all in UTC? ISO8601 allows for either.

EDIT: I agree with the general suggestion in Greg's comment to be very explicit when it comes to timestamps

---

File: specifications/cdsc-wg1-01.md [hunk 4]
```diff
@@ -1 +1,292 @@
...
+
+### 3.1. Metadata Well-Known URI <a id="metadata-well-known-uri" href="#metadata-well-known-uri" class="permalink">🔗</a>
+
+If the Server wishes to provide metadata publicly, the Server MUST support a metadata endpoint at the Well-Known URI `"/.well-known/carbon-data-spec.json"` on their hostname (e.g. https://example.com/.well-known/carbon-data-spec.json).
+
+Servers MAY additionally support other public and restricted URLs for special case or private metadata endpoints, as needed.
+
+### 3.2. Metadata Object Format <a id="metadata-object-format" href="#metadata-object-format" class="permalink">🔗</a>
+
+Metadata objects are formatted as JSON objects and contain named values, some of which are required and some of which are optional. The following values are included in the default list available in metadata objects.
```
Comment (daniel-utilityapi) [inline specifications/cdsc-wg1-01.md:74]: From @jhughes22 ([on May 25, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#discussion_r1205759610)):

---

Would a unique identifier of some kind for the Server entity make any sense? I see the form EIA-860 identifier used for this occasionally for US utilities in other contexts. In this case the Server does not have to be a utility, so it would probably need to be an optional field. Not a strong suggestion here, just something to think about. 

---

File: specifications/cdsc-wg1-01.md [hunk 5]
```diff
@@ -1 +1,292 @@
...
+* `map_content_type` - _MIME type_ - (OPTIONAL) Content-Type format of the `map_resource`
+* `geojson_resource` - _URL_ - (OPTIONAL) Link to a GeoJSON object mapping the coverage territory
+
+When `type` is a geographic entry type, one or both of `map_resource` and `geojson_resource` MUST be provided. If `map_resource` is provided, then `map_content_type` MUST also be provided.
+
+### 4.3. Coverage Entry Types <a id="coverage-entry-types" href="#coverage-entry-types" class="permalink">🔗</a>
+
+Coverage entry types describe what geographic territory is being detailed in the coverage entry. Many utilities have different geographic coverage for their different service types (e.g. their electric service territory is different from their natural gas service territory), so Servers may include multiple coverage entries for different types of service without having to publish multiple metadata endpoints.
+
+The following list of strings are an enumerated set of coverage entry types that are valid `type` values in the coverage entry object.
```
Comment (daniel-utilityapi) [inline specifications/cdsc-wg1-01.md:163]: From @jhughes22 ([on May 25, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#discussion_r1205771159)):

---

A balancing authority/ISO/TSO may have a coverage area that doesn't quite match any of these type descriptions. Something to consider if power systems related data is worked into the metadata advertising. The bit about extensions in section 6 may cover this though

---

Comment (daniel-utilityapi) [reply specifications/cdsc-wg1-01.md]: @jhughes22 Thanks for the feedback! I've moved infrastructure types to be in individual coverage entries and added `transmission_service_operator` and other types of infrastructure values to the most recent updates: https://daniel-utilityapi.github.io/Connectivity/specs/cdsc-wg1-01/#infrastructure-types

File: specifications/cdsc-wg1-01.md (hunk 4 above)
Comment (daniel-utilityapi) [reply specifications/cdsc-wg1-01.md]: Hmmm, I think this will be tricky since Servers are self-published, so there's not really any centralized naming authority to prevent collisions. The EIA-860 seems to be limited to only specific entity types, so maybe we could have an extension that allows for including that value as a capability?

File: specifications/cdsc-wg1-01.md (hunk 3 above)
Comment (daniel-utilityapi) [reply specifications/cdsc-wg1-01.md]: @jhughes22 Updated to specify `datetime` as RFC 3339, which requires a timezone offset.

File: specifications/cdsc-wg1-01.md (hunk 2 above)
Comment (daniel-utilityapi) [reply specifications/cdsc-wg1-01.md:137]: @halliecramer1 Thanks for the suggestion! Updated `infrastructure_type` to include `service_provider`. https://daniel-utilityapi.github.io/Connectivity/specs/cdsc-wg1-01/#infrastructure-types

**Comment by daniel-utilityapi** [comment]
> From @greg-flexidao ([on April 13, 2023](https://github.com/carbon-data-specification/Customer-Data/pull/23#issuecomment-1507149600)):
> 
> Hi All, Sorry I didn't yet check in depth the rest of the spec, my initial suggestions would be:
> 
>     * Consider date-time format according to RFC-3339 in order to explicitly define the time of the event/period. The example in the specification provides explicit time (ISO8601), note that it is also RFC-3339 compliant. ISO unfortunately also allows for ambiguous datetimes that are one of the main plagues and source of issues in the industry. See https://ijmacd.github.io/rfc3339-iso8601/ for comparison.
> 
>     * I would say that the endpoint/api should in general follow REST API principles. Unfortunately there is no ultimate source of truth for standard REST APIs (that I know of), here is however a comprehensive guide that could be followed and it can be referenced in the standard, that also includes the cases that have been mentioned. https://restfulapi.net/
> 
> 
> Happy to participate in further spec development!

@greg-flexidao Thanks for the feedback! I've updated the spec to use the date time format from RFC3339: https://daniel-utilityapi.github.io/Connectivity/specs/cdsc-wg1-01/#datetime

As far as REST principles, I agree that we do want to follow that best practice in general, but as you pointed out, there's not a clear specification we can use, so I think we should continue to just specify the individual endpoints.

File: specifications/cds-wg1-01.md [hunk 6]
```diff
@@ -1 +1,391 @@
...
+# CDS-WG1-01 - Server Metadata
+
+## Abstract <a id="abstract" href="#abstract" class="permalink">🔗</a>
+
+This specification defines how utilities and other entities may provide structured "metadata", such as reference information about them, functionality they offer, and how organizations can interoperate with them.
+The defined base metadata endpoint is intended to act as a starting point for organizations to discover details and capabilities of energy utilities and other similar entities.
+
+## Status <a id="status" href="#status" class="permalink">🔗</a>
+
+<b style="color:red">WARNING: This specification is a DRAFT and has not achieved consensus by the working group.</b>
```
Comment (daniel-utilityapi) [inline specifications/cds-wg1-01.md:10]: Update to note that this is a consensus draft, but not final.

**Comment by daniel-utilityapi** [comment]
I think this is ready to call for merging into the main CDS Registration WG website!

File: website/specs/cds-wg1-01/overview.md [hunk 7]
```diff
@@ -8,10 +8,115 @@ meta_description: Linux Foundation Energy - Connected Data Specifications (CDS)
...
+
+    "oauth_metadata": "https://demoutility.com/.well-known/oauth-authorization-server",
+
+    "ieee1547_setup": "https://demoutility.com/smart-inverters/setup"
+}
+```
+
+## Coverage Endpoint <a id="coverage-endpoint" href="#coverage-endpoint" class="permalink">🔗</a>
+
+This specification also defines a `coverage` capability that adds the the [Coverage Endpoint]({{ "/specs/cds-wg1-01" | relative_url }}#coverage-endpoint) for Servers to offer a way to list the covered geographic territories or logical groupings for which they offer the listed capabilities in a structured format.
```
Comment (eloifabrega) [inline website/specs/cds-wg1-01/overview.md:67]: typo

Comment (daniel-utilityapi) [reply website/specs/cds-wg1-01/overview.md]: @eloifabrega thanks for pointing this out! fixed in c00de88

File: website/specs/cds-wg1-01/overview.md [hunk 8]
```diff
@@ -8,10 +8,115 @@ meta_description: Linux Foundation Energy - Connected Data Specifications (CDS)
...
+                "electricity",
+            ],
+            "capabilities": [
+                "oauth",
+                "ieee1547"
+            ],
+
+            "map_resource": "https://static.example.com/cds-coverage-map.png",
+            "map_content_type": "image/png",
+            "geojson_resource": "https://static.example.com/cds-coverage-map-geo.json"
```
Comment (jhughes22) [inline website/specs/cds-wg1-01/overview.md:109]: A clarification on my comment in today's 07/10 meeting:

- Geojson defaults to WGS84 with longitude first, followed by latitude in the latest version of the RFC (2016). This should be unambiguous enough to use directly without additional verbiage. 
- Something to be aware of is that coordinate reference systems like EPSG:4326 use WGS84 datum but specify latitude first, longitude second.  So it can be a point of confusion anyways.
- Earlier versions of Geojson specifications allowed for customization of the CRS explicitly, causing interoperability issues and the CRS customization was intentionally removed. 

So, in short, I don't believe changes are needed with respect to my earlier feedback.

**Review by daniel-utilityapi — COMMENTED** [review]
Approved by working group to merge in 2025-08-07 meeting
